File: app/src/main/java/com/hikko/scheduleapp/adapters/WidgetService.java
package com.hikko.scheduleapp.adapters;

import android.appwidget.AppWidgetManager;
import android.content.Context;
import android.content.Intent;
import android.util.Log;
import android.view.View;
import android.widget.RemoteViews;
import android.widget.RemoteViewsService;

import com.hikko.scheduleapp.R;
import com.hikko.scheduleapp.Utils;

import java.util.ArrayList;
import java.util.HashMap;

public class WidgetService extends RemoteViewsService {

    @Override
    public RemoteViewsFactory onGetViewFactory(Intent intent) {
        return new WidgetViewsFactory(getApplicationContext(), intent);
    }



    static class WidgetViewsFactory implements RemoteViewsFactory {

        private final Context mContext;
        private static final String TAG = "WidgetService";
        private int appWidgetId;
        private ArrayList<HashMap<String, String>> activities;

        WidgetViewsFactory(Context context, Intent intent) {
            mContext = context;
            appWidgetId = intent.getIntExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, AppWidgetManager.INVALID_APPWIDGET_ID);
        }

        @Override
        public void onCreate() {
        }

        @Override
        public void onDataSetChanged() {
        }

        @Override
        public void onDestroy() {
        }

        @Override
        public int getCount() {
            activities = Utils.getActivitiesDayOfWeek(Utils.getLocaleDayOfWeek());
            Log.i(TAG, "activities " + activities);
            return activities.size();
        }

        @Override
        public RemoteViews getViewAt(int position) {
            RemoteViews views = new RemoteViews(mContext.getPackageName(), R.layout.widget_day_item);
            String name = activities.get(position).get("Name");
            views.setTextViewText(R.id.activity_start, activities.get(position).get("Start"));
            views.setTextViewText(R.id.activity_end, activities.get(position).get("End"));
            if (name == null || name.length() == 0) {
                View v = View.inflate(mContext, R.layout.activities_day_widget, null);
                views.setTextViewText(R.id.activity_name, v.getResources().getText(R.string.empty_name_activity));
                views.setTextViewText(R.id.activity_type, "");
            } else {
                views.setTextViewText(R.id.activity_name, name);
                views.setTextViewText(R.id.activity_type, activities.get(position).get("Type"));

            }
            Log.i(TAG, String.valueOf(views));
            return views;
        }

        @Override
        public RemoteViews getLoadingView() {
            return null;
        }

        @Override
        public int getViewTypeCount() {
            return 1;
        }

        @Override
        public long getItemId(int position) {
            return position;
        }

        @Override
        public boolean hasStableIds() {
            return true;
        }
    }

}

File: app/src/main/java/com/hikko/scheduleapp/adapters/WidgetService.kt
package com.hikko.scheduleapp.adapters

import android.appwidget.AppWidgetManager
import android.content.Context
import android.content.Intent
import android.view.View
import android.widget.RemoteViews
import android.widget.RemoteViewsService
import com.hikko.scheduleapp.Activity
import com.hikko.scheduleapp.ActivityUtils.getActivitiesDayOfWeek
import com.hikko.scheduleapp.ActivityUtils.localeDayOfWeek
import com.hikko.scheduleapp.R

class WidgetService : RemoteViewsService() {
    override fun onGetViewFactory(intent: Intent): RemoteViewsFactory {
        return WidgetViewsFactory(applicationContext, intent)
    }

    internal class WidgetViewsFactory(private val mContext: Context, intent: Intent) :
        RemoteViewsFactory {
        private val appWidgetId: Int
        private var activities: List<Activity>? = null

        init {
            appWidgetId = intent.getIntExtra(
                AppWidgetManager.EXTRA_APPWIDGET_ID,
                AppWidgetManager.INVALID_APPWIDGET_ID
            )
        }

        override fun onCreate() {}
        override fun onDataSetChanged() {}
        override fun onDestroy() {}
        override fun getCount(): Int {
            activities = getActivitiesDayOfWeek(localeDayOfWeek)
            return activities!!.size
        }

        override fun getViewAt(position: Int): RemoteViews {
            val views = RemoteViews(mContext.packageName, R.layout.widget_day_item)
            val name = activities!![position].name
            val cabinet = activities!![position].cabinet
            views.setTextViewText(R.id.activity_start, activities!![position].startTime)
            views.setTextViewText(R.id.activity_end, activities!![position].endTime)
            if (name == null || name.isEmpty()) {
                val v = View.inflate(mContext, R.layout.activities_day_widget, null)
                views.setTextViewText(
                    R.id.activity_name,
                    v.resources.getText(R.string.empty_name_activity)
                )
                views.setTextViewText(R.id.activity_type, "")
            } else {
                views.setTextViewText(R.id.activity_name, name)
                views.setTextViewText(R.id.activity_type, activities!![position].type)
                if (cabinet == null || cabinet.isEmpty()) {
                    views.setViewVisibility(R.id.activity_cabinet, View.GONE)
                } else {
                    views.setTextViewText(R.id.activity_cabinet, cabinet)
                    views.setViewVisibility(R.id.activity_cabinet, View.VISIBLE)
                }
            }
//            Log.i(TAG, views.toString())
            return views
        }

        override fun getLoadingView(): RemoteViews? {
            return null
        }

        override fun getViewTypeCount(): Int {
            return 1
        }

        override fun getItemId(position: Int): Long {
            return position.toLong()
        }

        override fun hasStableIds(): Boolean {
            return true
        }
    }
}